src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -5659,9 +5659,9 @@ int RGWRados::bucket_suspended(const DoutPrefixProvider *dpp, rgw_bucket& bucket
   return 0;
 }
 
-int RGWRados::Object::complete_atomic_modification(const DoutPrefixProvider *dpp, optional_yield y)
+int RGWRados::Object::complete_atomic_modification(const DoutPrefixProvider *dpp, bool keep_tail, optional_yield y)
 {
-  if ((!manifest)|| state->keep_tail)
+  if ((!manifest) || state->keep_tail || keep_tail)
```

Comment:
trying to understand where the regression happened

in `AppendObjectProcessor::prepare()` we conditionally set `astate->keep_tail = true;`. maybe that's a different instance of `astate` after zipper changes like https://github.com/ceph/ceph/pull/44616 or https://github.com/ceph/ceph/pull/57377?

passing through another `bool keep_tail` shouldn't be necessary if already have `state->keep_tail`

Reply:
keep_tail is set [here](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_rados.cc#L5015) in `RGWRados::copy_obj(..)`. However `astate` is for source object, not the [object](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_rados.cc#L4951) that the `write_meta` is called on, hence the flag is not brought in. 
Another solution for this is to populate the flag to the [object](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_rados.cc#L4951) manually before calling `write_meta`. The reason I didn't choose this one was because I was not 100% sure yet the `state` won't be over-written by later `get_obj_state`().
Which solution is better?

Reply:
> keep_tail is set [here](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_rados.cc#L5015) in `RGWRados::copy_obj(..)`. However `astate` is for source object, not the [object](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_rados.cc#L4951) that the `write_meta` is called on, hence the flag is not brought in.

hmm, maybe https://github.com/ceph/ceph/pull/58213 was the regression then?

Reply:
to avoid this confusion over different instances of `RGWObjectCtx`/`RGWObjState`, would it help to move `RGWObjState::keep_tail` to `RGWRados::Object::Write::MetaParams` to match our handling of `bool modify_tail`?

Reply:
> to avoid this confusion over different instances of `RGWObjectCtx`/`RGWObjState`, would it help to move `RGWObjState::keep_tail` to `RGWRados::Object::Write::MetaParams` to match our handling of `bool modify_tail`?

yup! thinking about that. I agree. Will make the change.

Reply:
done

Reply:
update looks good, but can `RGWObjState::keep_tail` be removed now?

Reply:
> update looks good, but can `RGWObjState::keep_tail` be removed now?

The `RGWObjState::keep_tail` is set [AppendObjectProcessor::prepare](https://github.com/ceph/ceph/blob/38d2c11bf24c347b2e27d17476528132d2bd0233/src/rgw/driver/rados/rgw_putobj_processor.cc#L720) as well. Is this one gonna be used later on? If it is, we either keep this field in `RGWObjState`, or change more places to make use of `RGWRados::Object::Write::MetaParams`.

I haven't traced the entire workflow of AppendObjectProcessor yet, and will do.
But if you know this on top of your head, please let me know so we can save some time.

Reply:
`AppendObjectProcessor::prepare()` sets `astate->keep_tail = true`, and then `AppendObjectProcessor::complete()` initializes the `RGWRados::Object::Write::MetaParams` for `write_meta()`. it looks like a `bool keep_tail` member variable could bridge that gap between prepare->complete

Reply:
> `AppendObjectProcessor::prepare()` sets `astate->keep_tail = true`, and then `AppendObjectProcessor::complete()` initializes the `RGWRados::Object::Write::MetaParams` for `write_meta()`. it looks like a `bool keep_tail` member variable could bridge that gap between prepare->complete

done